app/src/main/java/org/stepic/droid/code/ui/CodeAnalyzer.kt

```diff
@@ -0,0 +1,104 @@
+package org.stepic.droid.code.ui
+
+import org.stepic.droid.util.countWhile
+import org.stepic.droid.util.substringOrNull
+
+/**
+ * Class for smart code analyzing
+ */
+object CodeAnalyzer {
+    private const val LINE_BREAK = '\n'
+
+    private val brackets = hashMapOf(
+        "{" to "}",
+        "(" to ")",
+        "[" to "]"
+    )
+
+    private val indentation = hashMapOf( // code specific indentation symbols
+            "py" to hashSetOf(
+                    ":"
+            )
+    )
+
+    private val quotes = hashSetOf(
+        "\"", "'"
+    )
+
+    private fun getIndentForCurrentLine(start: Int, string: String) : Int {
+        val prevLineStart = string.substring(0, start).lastIndexOf(LINE_BREAK)
+        return string.countWhile (prevLineStart + 1) { Character.isWhitespace(it) && it != LINE_BREAK }
+    }
+
+    fun getIndentForLines(lines: List<String>) =
+            lines.map { getIndentForCurrentLine(0, it) }.filter { it > 0 }.min() ?: CodeEditor.DEFAULT_INDENT_SIZE
+
+    private fun getPrevSymbol(start: Int, string: String) : String? =
```

Comment:
Please, move it down (after using)

Reply:
Can we use char here?

Reply:
Now paired symbols are only 1 char length but in future, we can extend their sizes to any length. For example HTML tags.